lib/internal/modules/esm/loader.js

```diff
@@ -151,22 +151,28 @@ class DefaultModuleLoader {
     return this.getJobFromResolveResult(resolveResult, parentURL, importAssertions);
   }
 
-  getJobFromResolveResult(
+  async getJobFromResolveResult(
     resolveResult,
     parentURL,
     importAssertions = resolveResult.importAssertions,
   ) {
     const { url, format } = resolveResult;
-debug('getJobFromResolveResult:', { [url]: { format, moduleMap: this.moduleMap } });
+
+     // `job` may be a Promise, which the engine INSANELY _sometimes_ considers undefined and
+     // sometimes not. This is why we use `has` instead of `get` to check for its existence.
+     // ! Do NOT try to check against its value; the engine will **gladly** screw you over.
+    if (!this.moduleMap.has(url, importAssertions.type)) {
```

Comment:
i'm confused why a map's "has" function would ever take more than one argument. is `moduleMap` something that's not a Map?

Reply:
Good question! I also had to ask what this is.

The extra argument (import assertion type) is per spec required for matching cache. Under the hood, the extra argument is stored as part of the value (or maybe the key?).

I'll add some code doc for it after I get this stable/passing.

Reply:
I think it’s very confusing to have something called “map” whose method names violate the expected interface contract for a Map, but i assume that predates this PR. (iow, i think instead of using has/set/etc it should have brand new method names)

Reply:
Yes, this PR does not introduce `ModuleMap` ;)